### ISSUE TITLE
Best way to pass pre-trained image embeddings into model

### ISSUE THREAD

**Issue by nialloh23** [tensorflow/recommenders#175]
I'm trying to incorporate pre-trained image embeddings into my ranking model.
I'll outline some high level details on my approach below and then 2 short questions I had on the implementation:

**1. Building the tf dataset:**
I start with a datframe of style_ids(int), user_ids(int) and image_embeddings(np arrays).
```
        style = np.array([[i] for i in self.df_interactions['style'].values])
        user_id = np.array([[i] for i in self.df_interactions['user_id'].values])
        image_embedding = np.array([i for i in self.df_interactions['embedding'].values])

        interactions_tf_dataset = {'user_id': user_id,
                                   'style': style,
                                   'image_embedding':image_embedding}
     interactions = tf.data.Dataset.from_tensor_slices(interactions_tf_dataset)
```

**2. Concatenating Embeddings with Pre-trained Image Embeddings**
I'm setting up the user and style embeddings as normal but this time I'm concatenating pre-trained image embeddings directly in the call. The concatenated embeddings are of shape: (?, 1, 64) (?, 1, 64) and (?, 1,64)

```
  def call(self, inputs):
    user_embedding = self.user_embeddings(inputs['user_id'])
    style_embedding = self.style_embeddings(inputs['style_id'])
    image_embeddings = tf.expand_dims(features['image_embedding'], axis=1)

    concatenated_embeddings = tf.concat([user_embedding, style_embedding, image_embeddings], axis=1))
```

**Questions:**
1. To be able to concatenate the image embeddings with the user, style embeddings I compressed them down to the same dimension (i.e. I kept all to 64 dim). Do I need to do this? Or can I keep my image embeddings at say 1000 dim and concatenate along a different access? (eg. axis=-1, axis=2)? 

2. My training appears more unstable since adding the image embeddings? I was wondering is it best practice to normalize these embeddings in some way before passing in(they are currently normalized to be in the range -1 to 1)? Or is there a mismatch between initialized style/id embeddings and the input image embedding vectors which are fixed?

Thanks in advance for any help/pointers!

**Comment by maciejkula** [comment]
1. You should probably get rid of the 1-sized dimension in your image embeddings to make them of shape `(?, 64)`, then concatenate along the last dimension. This will allow you to choose any dimensions you like. In general, most of the tensors you operate on will be 2D unless you're dealing with sequences or raw images.
2. Once you have both embeddings and image features, you should probably put some dense layers on top - this will help scale them. You can also try normalizing the embeddings to unit length.
3. You can always make the image embedding model part of your TFRS model: this will allow you to pass raw images as inputs to your model.

**Comment by nialloh23** [comment]
Thanks @maciejkula. Your suggestions worked.

1. I 'm now concatenating along the last dimension which means I can pre-trained image embeddings which are different in length to my feature embeddings.
2. I also needed to fix the way I was craeting my tf_dataset. I was making everything into numpy arrays which was unneccesary and causing issues with shape match on concatenation.

For anyone who comes on this again here is some sample code from a working solution:
```
        style = self.df_interactions['style'].values
        user_id = self.df_interactions['user_id'].values
        image_embedding = np.array([i for i in self.df_interactions['embedding'].values])

        interactions_tf_dataset = {'user_id': user_id,
                                   'style': style,
                                   'image_embedding':image_embedding}
     interactions = tf.data.Dataset.from_tensor_slices(interactions_tf_dataset)
```
And within my model:
```
  def call(self, inputs):
    user_embedding = self.user_embeddings(inputs['user_id'])
    style_embedding = self.style_embeddings(inputs['style_id'])
    image_embeddings = tf.expand_dims(features['image_embedding'], axis=1)

    concatenated_embeddings = tf.concat([user_embedding, style_embedding, image_embeddings], axis=-11))
```

**Comment by nialloh23** [comment]
This question has been resolved. Thanks @maciejkula
